dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorLitmusAArch64.java

```diff
@@ -105,163 +143,302 @@ public Object visitInstructionRow(LitmusAArch64Parser.InstructionRowContext ctx)
             throw new ParsingException("Comparing non-integer register.");
         }
         IntLiteral zero = expressions.makeZero(integerType);
         Expression expr = expressions.makeIntCmp(register, ctx.branchRegInstruction().op, zero);
         Label label = programBuilder.getOrCreateLabel(mainThread, ctx.label().getText());
-        return programBuilder.addChild(mainThread, EventFactory.newJump(expr, label));
+        return add(EventFactory.newJump(expr, label));
     }
 
     @Override
-    public Object visitBranchLabel(LitmusAArch64Parser.BranchLabelContext ctx) {
-        return programBuilder.addChild(mainThread, programBuilder.getOrCreateLabel(mainThread, ctx.label().getText()));
+    public Object visitBranchLabel(BranchLabelContext ctx) {
+        return add(programBuilder.getOrCreateLabel(mainThread, ctx.label().getText()));
     }
 
     @Override
-    public Object visitFence(LitmusAArch64Parser.FenceContext ctx) {
-        return programBuilder.addChild(mainThread, EventFactory.newFenceOpt(ctx.Fence().getText(), ctx.opt));
+    public Object visitFence(FenceContext ctx) {
+        return add(EventFactory.newFenceOpt(ctx.Fence().getText(), ctx.opt));
     }
 
     @Override
-    public Object visitReturn(LitmusAArch64Parser.ReturnContext ctx) {
+    public Object visitReturn(ReturnContext ctx) {
         Label end = programBuilder.getEndOfThreadLabel(mainThread);
-        return programBuilder.addChild(mainThread, EventFactory.newGoto(end));
+        return add(EventFactory.newGoto(end));
     }
 
     @Override
-    public Expression visitExpressionRegister64(LitmusAArch64Parser.ExpressionRegister64Context ctx) {
+    public Expression visitExpressionRegister64(ExpressionRegister64Context ctx) {
         Expression expr = programBuilder.getOrNewRegister(mainThread, ctx.register64().id, archType);
         if(ctx.shift() != null){
-            IntLiteral val = expressions.parseValue(ctx.shift().immediate().constant().getText(), archType);
+            IntLiteral val = parseValue(ctx.shift().immediate().constant(), archType);
             expr = expressions.makeIntBinary(expr, ctx.shift().shiftOperator().op, val);
         }
         return expr;
     }
 
     @Override
-    public Expression visitExpressionRegister32(LitmusAArch64Parser.ExpressionRegister32Context ctx) {
-        Expression expr = programBuilder.getOrNewRegister(mainThread, ctx.register32().id, archType);
-        if(ctx.shift() != null){
-            IntLiteral val = expressions.parseValue(ctx.shift().immediate().constant().getText(), archType);
-            expr = expressions.makeIntBinary(expr, ctx.shift().shiftOperator().op, val);
+    public Expression visitExpressionRegister32(ExpressionRegister32Context ctx) {
+        final Register r64 = programBuilder.getOrNewRegister(mainThread, ctx.register32().id, archType);
+        final Expression expr = expressions.makeIntExtract(r64, 0, 31);
```

Comment:
OMG! Why are we creating a i64 register to then extract the 32 bits we care about?

Reply:
I think this is because `W?` refers to the lower 32 bits of the 64 bit `X?` register, no?
In particular, all general purpose registers are always 64 bit, so this modeling looks accurate to me.